### PR TITLE
[FIRRTL][LowerClasses] Lower map types and map.create

### PR DESCRIPTION
**Pull request by uenoku**
This implements Map type lowerings. Currently reject non-string or integer key types in type converter.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 1]
```diff
@@ -437,6 +457,32 @@ static void populateTypeConverter(TypeConverter &converter) {
         return om::ListType::get(elementType);
       });
 
+  auto convertMapType = [&converter](auto type) -> std::optional<mlir::Type> {
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:460]: Good call re:lambda, should do same for the list conversions too probably (not in this PR, something I can touchup) :+1: .

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 2]
```diff
@@ -323,6 +323,26 @@ struct ListCreateOpConversion
...
+                  ConversionPatternRewriter &rewriter) const override {
+    auto mapType = getTypeConverter()->convertType<om::MapType>(op.getType());
+    if (!mapType)
+      return failure();
+    auto keys = adaptor.getKeys();
+    auto values = adaptor.getValues();
+    SmallVector<Value> tuples;
+    for (auto [key, value] : llvm::zip(keys, values))
+      tuples.push_back(rewriter.create<om::TupleCreateOp>(
+          op.getLoc(), ArrayRef<Value>{key, value}));
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:340]: Could this be simply `{key, value}`? Not important, and if you prefer this regardless works for me.

File: test/Dialect/FIRRTL/lower-classes.mlir [hunk 3]
```diff
@@ -89,4 +89,45 @@ firrtl.circuit "Component" {
     %true = firrtl.bool true
     firrtl.propassign %b, %true : !firrtl.bool
   }
+
+  // CHECK-LABEL: om.class @MapTest
+  firrtl.class @MapTest(in %s1: !firrtl.string,
+                         in %s2: !firrtl.string,
```
Comment (dtzSiFive) [inline test/Dialect/FIRRTL/lower-classes.mlir:95]: Nit: alignment, looks like args are one column too far to the right.

**Review by dtzSiFive — APPROVED** [review]
LGTM, added some minor comments/suggestions.

Please add at least one error test, especially for the handling of unsupported key types.

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:340]: Tried  `{key, value}` but didn't work without explicit types 😞

Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:340]: Okay, thanks!

**Pull request merged**